# Import Necessary Libraries

In [6]:
import numpy as np
from sympy import Matrix

# 1. Create Your Personal Matrix

In [5]:
d1 = 5 #last digit of id
d2 = 8 #second last digit of id

A = np.array([[d1+2, d2+1], [2*d1, d2+2]])

print(A)

[[ 7  9]
 [10 10]]


# 2. Analyze Matrix A

In [7]:

# 1. Shape of A
shape = A.shape
print(f"Shape of A: {shape}")

# 2. Determinant
det = np.linalg.det(A)
print(f"Determinant: {det:.2f}")

# 3. Rank
rank = np.linalg.matrix_rank(A)
print(f"Rank: {rank}")

# 4. Eigenvalues
eigenvalues = np.linalg.eigvals(A)
print(f"Eigenvalues: {eigenvalues}")

# 5. Inverse (only if determinant != 0)
if det != 0:
    inv = np.linalg.inv(A)
    print("Inverse of A:")
    print(inv)
else:
    print("Matrix A is singular and does not have an inverse.")

Shape of A: (2, 2)
Determinant: -20.00
Rank: 2
Eigenvalues: [-1.10468636 18.10468636]
Inverse of A:
[[-0.5   0.45]
 [ 0.5  -0.35]]


# 3. Value Change Experiment

In [8]:
# Creating matrix B by changing one value in A
# Choice: Add +1 to element A[0, 0]

B = A.copy()
B[0, 0] += 1

print("Original Matrix A:")
print(A)
print("\nNew Matrix B (A[0,0] + 1):")
print(B)

print(f"\nChanged element at index (0,0) from {A[0,0]} to {B[0,0]}")

Original Matrix A:
[[ 7  9]
 [10 10]]

New Matrix B (A[0,0] + 1):
[[ 8  9]
 [10 10]]

Changed element at index (0,0) from 7 to 8


# 4. Re-analyze Matrix B

In [9]:
# Re-analyze Matrix B

# 1. Shape of B
shape_B = B.shape
print(f"Shape of B: {shape_B}")

# 2. Determinant
det_B = np.linalg.det(B)
print(f"Determinant of B: {det_B:.2f}")

# 3. Rank
rank_B = np.linalg.matrix_rank(B)
print(f"Rank of B: {rank_B}")

# 4. Eigenvalues
eigenvalues_B = np.linalg.eigvals(B)
print(f"Eigenvalues of B: {eigenvalues_B}")

# 5. Inverse (only if determinant != 0)
if not np.isclose(det_B, 0):
    inv_B = np.linalg.inv(B)
    print("Inverse of B:")
    print(inv_B)
else:
    print("Matrix B is singular and does not have an inverse.")

Shape of B: (2, 2)
Determinant of B: -10.00
Rank of B: 2
Eigenvalues of B: [-0.53939201 18.53939201]
Inverse of B:
[[-1.   0.9]
 [ 1.  -0.8]]


# 5. Comparison & Explanation

### 1. How did the determinant change and why?
The determinant changed from **-20.00** in Matrix A to **-10.00** in Matrix B. This change occurred because I increased the element at index [0, 0] by 1. For a 2x2 matrix, the determinant is calculated as $(a_{11} \times a_{22}) - (a_{12} \times a_{21})$. By increasing $a_{11}$ from 7 to 8 while keeping $a_{22}$ at 10, the first product increased by 10 ($1 \times 10$), which mathematically explains the shift from -20 to -10.

### 2. Did the rank change? Explain.
No, the **rank did not change**; it remained **2** for both matrices. This is because in both cases, the rows (and columns) are linearly independent. Since the determinant is non-zero for both A and B, they are both full-rank matrices.

### 3. How did the eigenvalues respond to the value change?
The eigenvalues shifted slightly. In Matrix A, they were approximately **-1.10** and **18.10**. In Matrix B, they became approximately **-0.54** and **18.54**. Increasing the value of a diagonal element generally increases the sum of the eigenvalues (the trace of the matrix), which is consistent with our observation.

### 4. Is B easier or harder to invert than A? Why?
Mathematically, B is **slightly easier** to invert than A because its determinant is further away from zero in relative terms, or more precisely, the resulting inverse values are "cleaner" integers/fractions. In Matrix A, the inverse elements were -0.5 and 0.45, while in Matrix B, they became -1.0, 0.9, 1.0, and -0.8. A determinant closer to zero generally makes a matrix more numerically unstable (closer to being singular), so increasing the determinant magnitude (from -20 to -10 in this specific context of inverse scaling) changed the scaling factor of the adjoint matrix.